rfcs/20190506-filesystem-plugin-modular-tensorflow.md

```diff
@@ -0,0 +1,1082 @@
+   implement a filesystem functionality. Each time the boundary is crossed we
+   have to do conversions from C++ types to C types and back and that incurs
+   some performance costs.
+
+1. The API should provide some _type safety_, to ensure that plugin operations
+   only act upon data structures that they should work on. Thus, instead of
+   using opaque `void*` pointers, we will need to differentiate between the
+   structures being operated on. We do this by having data structures which just
+   wrap around these opaque pointers. These structures will only have one
+   `void*` field.
+
+1. We should provide a way for plugins to register that they don’t support some
+   functionality, either because it doesn’t make sense for the specific plugin
+   or because of compatibility reasons.
+
+1. We should have a minimal amount of allocation at the boundary between plugins
+   and core TensorFlow. Most of the memory should either be owned by TensorFlow
+   or by the plugin and the only transfer between these should happen when
+   reading/writing files as well as in converting from C++ types to C types
+   (e.g. marshalling from `string` to `char*`)
+
+1. The proposed API should minimize the amount of sweeping changes that need to
+   happen on TensorFlow code. Ideally, we should allow some period of time when
+   both plugins and the existing filesystem implementation will be present in
+   the code, allowing users to gradually shift towards the new design.
+   Fortunately, the proposed design allows for transparently rewriting the
+   current C++ APIs, causing minimal disruption.
+
+1. Although this design is mostly about C++ implementation, we should ensure
+   Python code and other language bindings will be able to continue working with
+   the modular design.
+
+## Existing implementation
+
+The existing implementation of a filesystem support in TensorFlow is hairy, due
+to the complexity of operations that needs to be provided. This section is an
+overview of the existing architecture.
+
+In short summary, TensorFlow’s support for filesystems can be decomposed into 3
+types of API:
+
+* Low level API mainly focused on offering interfaces to hide complexity and
+  provide cross-platform compatibility;
+* Convenience API providing functions to read/write files in a generic way as
+  well as providing support for compressed data and for buffered I/O;
+* High level API which offers functionality needed by the kernels, ops and other
+  utilities.
+
+Furthermore, the codebase has some additional methods to access the filesystem.
+These are used sporadically in some kernels and ops or for special cases such as
+very special filesystems or test environments. We can translate these additional
+APIs to use the design presented here, unifying API all across the codebase.
+
+![Filesystem layers](20190502-filesystem-plugin-modular-tensorflow/big_picture.png "Overview of filesystem support")
```

Comment:
Image path wrong, the date in the folder has changed.

Reply:
Fixed now, thanks